internal/datastore/mysql/readwrite.go

```diff
@@ -0,0 +1,288 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/jzelinskie/stringz"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/internal/datastore/mysql/migrations"
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+const (
+	errUnableToWriteRelationships  = "unable to write relationships: %w"
+	errUnableToDeleteRelationships = "unable to delete relationships: %w"
+	errUnableToWriteConfig         = "unable to write namespace config: %w"
+	errUnableToDeleteConfig        = "unable to delete namespace config: %w"
+)
+
+type mysqlReadWriteTXN struct {
+	*mysqlReader
+
+	ctx      context.Context
+	tx       *sql.Tx
+	newTxnID uint64
+}
+
+// WriteRelationships takes a list of existing relationships that must exist, and a list of
+// tuple mutations and applies it to the datastore for the specified namespace.
+func (rwt *mysqlReadWriteTXN) WriteRelationships(mutations []*v1.RelationshipUpdate) error {
+	// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+	// there are some fundamental changes introduced to prevent a deadlock in MySQL
+	ctx, span := tracer.Start(datastore.SeparateContextWithTracing(rwt.ctx), "WriteTuples")
+	defer span.End()
+
+	bulkWrite := rwt.WriteTupleQuery
+	bulkWriteHasValues := false
+
+	selectForUpdateQuery := rwt.QueryTupleIdsQuery
+
+	clauses := sq.Or{}
+
+	// Process the actual updates
+	for _, mut := range mutations {
+		rel := mut.Relationship
+
+		// Implementation for TOUCH deviates from PostgreSQL datastore to prevent a deadlock in MySQL
```

Comment:
@vroldanbet Can you explain more context here on how this prevents a deadlock? We might want to extend the comment here

Reply:
I'm going to defer to @bryanhuhta as they worked on the actual solution back in the day. For what's worth, the changes decreased the likelihood of the dead-lock, but it's still happening when doing massive concurrent imports.

Reply:
Actually maybe we want to allow it to deadlock, that should be faster to detect and retry than the current lock wait timeout that we're doing.

Reply:
Allowing the deadlocks to happen could be faster, I haven't really measured the impact of a deadlock detection and rollback.

> Can you explain more context here on how this prevents a deadlock? We might want to extend the comment here

The TL;DR of the situation is MySQL takes greedy row locks when trying to update a row that doesn't exist, whereas Postgres doesn't have the same behavior.

The longer version: With the Postgres implementation, there are basically two database actions that happen in a transaction during a `TOUCH` operation (assuming a single tuple):

- Update the most recent tuple record as "deleted"
- Insert a new tuple record as "newest"

Using that same logic in MySQL, two locks would need to be gathered:

- **next-key lock** to block writes on the row (multiple transactions can hold this at once)
- **insert intention gap lock** to insert a new record (multiple transactions can hold this, as long as they aren't inserting into the same gap)

This is fine if the tuple to marked as "deleted" already exists. However, if it doesn't, then the next-key lock will lock rows between the last row and the `supremum` row. This creates a situation where transaction `T1` and `T2` are trying to perform a `TOUCH` on two records that don't exist:

```
*
| T1 acquires next-key lock (locks rows 1 to supremum)
| T2 acquires next-key lock (locks rows 1 to supremum) 
| T1 acquires gap lock (locks row of next possible id, which is between 1 and supremum)
| T2 acquires gap lock (locks row of T1 + 1, which is between 1 and supremum)
!! Deadlock, T1 needs to insert into a gap that T2 holds a lock on (via next-key lock)
!! and T2 needs to insert into a gap that T1 holds a lock on (via next-key lock). T2 rolls back.
```

In our case, the fix was to only update rows that exist in the database. That stops a transaction from acquiring a next-key lock up to `supremum`. We still have deadlocks in certain situations that I haven't quite been able to figure out yet.

For a better explanation of how this deadlock occurs, there's a 2 million year old MySQL bug report: https://bugs.mysql.com/bug.php?id=25847.

Reply:
> there's a 2 million year old MySQL bug report

😂 

The bug mentions using `select for update`, and the MySQL code calls the query `selectForUpdateQuery`, but I don't actually see where any queries are actually using the `for update` modifier. Could this be related to your ongoing deadlocks?

I would like to work together to get real data on whether it's faster/better to just let the deadlock happen and retry the transaction. This is more or less the entire justification for doing the datastore v2 work now. We have similar client-side retry requirements across all of our datastores.